Performance, memory and styling improvements in OpFastScintillation

ikatza opened the pull request:
Have done some major refactoring of the code inside `LegacyLArG4/OpFastScintillation`. 

Testing on a sbndgpvm on Cosmics + Neutrino events; the most cpu and memory hungry of events, I have achieved close to half of running time and a comparable memory footprint.

Despite using some fast approximations to the most used functions there is no measurable bias in the OpHits.

Along the way I have also address some of the underlying potential bugs that were hidden in the code. The most salient one had to with  a domain error exception calling the boost elliptic integral functions, as far as I can tell this would happen due to a floating point error bloating.

ikatza committed:
Styling and whitespace removal.

Used the script:
astyle --style=stroustrup --indent=spaces=2  --indent-namespaces --lineend=linux --pad-oper --pad-comma --indent-col1-comments --keep-one-line-statements OpFastScintillation.cxx  OpFastScintillation.hh

and manually trim some lines.

ikatza committed:
Remove test. Boost already checks the purported objective of this test.

ikatza committed:
Use std everywhere

ikatza committed:
Avoid using pow on Pol_5. Return a regular double.

ikatza committed:
Don't spend extra cycles on std::sqrt(std::pow()), use std::abs instead.

ikatza committed:
Create function dist() instead of using std::pow.

In C++11 std::pow casts the exponent to double. This approach saves memory.

ikatza committed:
Deobfuscate the recusivity of Disk_SolidAngle() to a single function.

ikatza committed:
Define function to make comparisons that take floating precission and tolerance into account, use them in Disk_SolidAngle()

I noticed that the equal condition was hardly being satisfied. Now it should be more often.
This might change results somewhat.

ikatza committed:
Remove expression that is identical to zero, simplify the other one.

ikatza committed:
Change interpolate() to be void to stop returning a dynamically allocated pointer that was created inside a function.

Bit surprised the compiler didn't prevent this.

ikatza committed:
Stop allocating and returning vectors, instead use a single vector that gets passed along by ref.

ikatza committed:
Prevent boost from promoting to long double, this substantially improves performance but might introduce some bias.

ikatza committed:
Remove gratuitous pow

ikatza committed:
Refactor time smearing loop

ikatza committed:
Correct scope of vars

ikatza committed:
Swap ints to size_t and const where appropriate

ikatza committed:
Split long lines, collapse a few others

ikatza committed:
Improve encapsulation and compiling times by moving functions around

ikatza committed:
Reorder ifs in Rectangle_SolidAngle in decreasing regularity of usage

ikatza committed:
Include function to generate table of acos

ikatza committed:
Include function to compute acos quickly at the expense of precision

ikatza committed:
Mark collection of improvements to be made

ikatza committed:
Define and use comparison of floating point variable to zero

ikatza committed:
WIP function to create memoized functions

ikatza committed:
Stop evaluating and copying constant structs values. The struct now only has what it needs.

ikatza committed:
Choose fast_acos to handle computations, substantial performance improvement at the expense of precision

ikatza committed:
Offload some of the calculations to the compiler marking functions as constexpr

ikatza committed:
Use std functions on fast_acos

ikatza committed:
Use std functions and CLHEP::pi to be able to use constexpr in Disk_SolidAngle, also better consistency

ikatza committed:
Move orders of ifs in Disk_SolidAngle

Note that in testing it only landed on the first two options.

ikatza committed:
Refactor Rectangle_SolidAngle to reduce abs calls and memory usage

ikatza committed:
Change type of parameter of Rectangle_SolidAngle to std::array, economise on declarations and reorder expressions to prevent cache misses

ikatza committed:
Remove alias variable and use the one that was passed along to Rectangle_SolidAngle()

ikatza committed:
Revist order of ifs in Rectangle_SolidAngle, make this one a constexpr, some more tweaking.

ikatza committed:
Merge branch 'develop' into feature/icaza_performance

Handled conflict manually and with care to put the same logic.

ikatza committed:
Skip the volume in between the cathode planes

ikatza committed:
Annotate a potential improvement

ikatza committed:
Merge branch 'feature/icaza_performance' into develop

FNALbuild commented:
A new Pull Request was created by  @ikatza (Iker de Icaza) for develop.

It involves the following packages:

larsim


@LArSoft/level-1-managers, @LArSoft/level-2-managers can you please review it and eventually sign? Thanks.

cms-bot commands are listed <a href="https://cdcvs.fnal.gov/redmine/projects/larsoft/wiki/Pull_request_comments_that_trigger_CI_actions">here</a>

FNALbuild commented:
The code-checks are being triggered in jenkins.

FNALbuild commented:
+code-checks  
 for details see 
 https://buildmaster.fnal.gov/buildmaster/job/GithubIntegration/job/run-pr-code-checks//500/console

aihimmel commented:
Would it be possible to also merge this work into the new LArG4? I worry about major development happening in the Legacy simulation code and being lost when we switch over to the new LArG4.

ikatza commented:
Yeah I can do it for sure. But I would rather make another PR for that.

aihimmel commented:
No problem. I've read through the changes. They look dramatic, but it looks like the vast majority were badly needed re-styling, plus removing some equality checks on floats (very nice). But, it looks like this also includes optimizations which will make minor changes in results. Can you share some more of the details of the checks you did with SBND simulation? Small changes (rather than identical results) make validation somewhat more challenging.

ikatza commented:
I don't think these changes can be validated with the CI system, there are a number of RNG inside without reproducible seeds.

The solid angle functions are run many times, these are still where most cycles are spend.

To improve the Disk_SolidAngle function I ask boost to not promote doubles to long doubles. Boost internally promotes to make sure the errors are small. That's on: d158b9c

To improve Rectangle_SolidAngle I implemented a fast arc cosine function, that's on: 16efd72 and 09e5951. 

I ran this commit:
https://github.com/ikatza/larsim/commit/26fe8341bd3f00d4ec535321824292f30f0241a3
(note it's not on this PR) to make sure that the lack of precision in those changes is not greater than 1%. I've run this with about 20 cosmics + BNB events with no warning being printed out.

Given that `solid_angle` is then put into a Poisson function and then rounded up, I believe 1% accuracy is more than enough.

Also had an attempt at making tables of the ellint and acos full precision functions, but that prove way trickier to make better. I left the code there in case it can be picked up in the future. 

I made a similar test to Pol_5() function, but didn't save it.

aihimmel commented:
Yea, that was my point. CI can't check this since identity is not expected. Have you, for example, simulated a few k events before and after your changes and made some distributions to make sure things look the same statistically (even if they vary on individual events)?

And, speaking of CI, in the DUNE CI we're seeing some changes in data products beyond just the PD-related ones:
```
2106: 10c10
2107: < G4 | largeant |  | std::vector<sim::OpDetBacktrackerRecord> | 366
2108: ---
2109: > G4 | largeant |  | std::vector<sim::OpDetBacktrackerRecord> | 419
2110: 13c13
2111: < G4 | largeant |  | std::vector<simb::MCParticle> | 884
2112: ---
2113: > G4 | largeant |  | std::vector<simb::MCParticle> | 384
2114: 15,19c15,19
2115: < G4 | largeant | Reflected | std::vector<sim::OpDetBacktrackerRecord> | 479
2116: < G4 | largeant |  | art::Assns<simb::MCTruth,simb::MCParticle,sim::GeneratedParticleInfo> | 884
2117: < G4 | largeant | Reflected | std::vector<sim::SimPhotonsLite> | 479
2118: < G4 | largeant |  | std::vector<sim::SimChannel> | 2842
2119: < G4 | largeant |  | std::vector<sim::SimPhotonsLite> | 366
2120: ---
2121: > G4 | largeant | Reflected | std::vector<sim::OpDetBacktrackerRecord> | 480
2122: > G4 | largeant |  | art::Assns<simb::MCTruth,simb::MCParticle,sim::GeneratedParticleInfo> | 384
2123: > G4 | largeant | Reflected | std::vector<sim::SimPhotonsLite> | 480
2124: > G4 | largeant |  | std::vector<sim::SimChannel> | 3132
2125: > G4 | largeant |  | std::vector<sim::SimPhotonsLite> | 419
```

Could these random-number-related changes be the source of the changes in MCParticles and SimChannels?

ikatza commented:
Had a go at that but didn't got far as I assumed it wouldn't be able to tell a difference: 

Made a 100 muon gun events, same momentum, position and direction in all of them. The standard deviation of the total number of photons generated was half the mean. 

Frankly I fail to understand how to make a test for this. If you could help me out with that I'm happy to set it up.


I see some crazy variation on the data product sizes, sorry no idea why it changes so wildly and beyond PD.

aihimmel commented:
I think you've got the right idea, though 100 muons all in the same place is maybe not a sufficient test given that this is all about how the light behaves in different location. Perhaps you could do a few hundred, say 500 MeV electrons (more localized) at various locations around the volume, and then compare some distributions in how much light you see on them from different channels. 

As I said -- validating changes which are expected to be the same statistically, but not identical event by event, are very challenging.

aihimmel commented:
Sorry -- apparently the mysterious change in SimChannels was from a different pull request. But I think seeing some form of validation here would still be wise.

ikatza committed:
Remove gratuitous else

ikatza committed:
Stop creating ScintPoint and OpDetPoint at every iteration of the loops. Use simpler const std::array.

dist() function changed accordingly. Similarly with fOpDetCenter.

ikatza committed:
Move computation of reflected hits to outside the loop as it doesn't depend on OpDetPoint.

ikatza committed:
Encapsulate conditions for computing semi-analythic and call them before VISHits and VUVHits

ikatza committed:
Consistency with literal doubles

knoepfel commented:
@ikatza, are you able to construct a test along the lines Alex has requested?

ikatza commented:
Yeah I can. I'm discussing how to do it. I figured this test could help other experiments as well and for transitioning to the new LArG4.

It will take me at least one week, as I have some other projects to catch up on.

aihimmel commented:
Thanks Iker. Sorry to slow things down, but this is the sort of change which has the danger of introducing very difficult to track down bugs (which we have had some experience with already), so it's much better to check it carefully up front.

ikatza committed:
Put a try/catch block on Disk_SolidAngle for the rare cases where an exception might be thrown

FNALbuild commented:
Pull request #14 was updated. @LArSoft/level-1-managers, @LArSoft/level-2-managers can you please check and sign again.

FNALbuild commented:
The code-checks are being triggered in jenkins.

FNALbuild commented:
+code-checks  
 for details see 
 https://buildmaster.fnal.gov/buildmaster/job/GithubIntegration/job/run-pr-code-checks//505/console

ikatza commented:
FYI I came up with some more improvements. Processing time down to 42% of vanilla OpFastScintillation.

therealendrju commented:
Hi Iker, these look like significant improvements. While I agree that the running time speed up is welcome and indeed crucial,  we (SBND) would like to see the tests Alex mentioned and a breakdown of what the changes are as well as their effects on precision etc.

ikatza committed:
Instead of dividing multiply by reciprocal in Rectangle_SolidAngle

ikatza committed:
Refactor interpolate functions.

Avoid cache misses, reuse idx, make it part of OpFastScintillation
in the hopes it becomes inlined, etc.

ikatza committed:
Prevent gratuitous casting

ikatza committed:
Encapsulate and distinguish in two distinct functions the code that gets direct and reflected hits

ikatza committed:
Acted on some TODOs, chatted with Diego about them

ikatza committed:
Use std::acos() in places where it's not called often

ikatza committed:
Temporary change to reproduce vanilla OpFastScintillation

FNALbuild commented:
Pull request #14 was updated. @LArSoft/level-1-managers, @LArSoft/level-2-managers can you please check and sign again.

FNALbuild commented:
The code-checks are being triggered in jenkins.

FNALbuild commented:
+code-checks  
 for details see 
 https://buildmaster.fnal.gov/buildmaster/job/GithubIntegration/job/run-pr-code-checks//533/console

ikatza commented:
This new round of changes comes with some additional performance and memory improvements.

Running standard_larg4 on a gpvm for BNB+Cosmics events I measure a performance improvement of 2.6 x faster with my changes. There are similar performance improvements, but not as good, for less busy events. 

I have produced a statistical test in the framework of SBND with the changes I have just pushed.

![image](https://user-images.githubusercontent.com/8713617/81322549-11ff8100-908c-11ea-93d0-64ee6cdcbb59.png)


Looking at a collection of PMTs that represent different positions on SBND I get:
![image](https://user-images.githubusercontent.com/8713617/81322701-42471f80-908c-11ea-895b-760d3c053b9b.png)

then for the percentage difference in some of those channels and at the bottom the percentage difference of all photons detected by PMTs:
![image](https://user-images.githubusercontent.com/8713617/81322854-76224500-908c-11ea-9fe8-80e904cbe1a8.png)


Similarly for ARAPUCAS
![image](https://user-images.githubusercontent.com/8713617/81322896-833f3400-908c-11ea-8bdb-b656c0494a6c.png)

then for the percentage difference in some of those channels and at the bottom the percentage difference of all photons detected by ARAPUCAS
![image](https://user-images.githubusercontent.com/8713617/81322962-981bc780-908c-11ea-8bd9-38c76370398b.png)


When looking at all the photons collected in the specific TPC where I produced the events I get the percentage differences of the two components of light:
![image](https://user-images.githubusercontent.com/8713617/81323103-cac5c000-908c-11ea-8af5-8d1e1951fbb2.png)


Variations of ~1% are expected due to the random sampling not being systematically seeded everywhere. Here I show the differences between two equivalent runs:
![image](https://user-images.githubusercontent.com/8713617/81323223-f648aa80-908c-11ea-9211-9981a1598c8c.png)

aihimmel commented:
Iker, thanks for doing these careful checks. I'm happy to see that you can reproduce the previous simulation behavior with even better speed-ups! Looking forward to the next PR with the bug fixes included as well.

aihimmel commented:
Approve

FNALbuild commented:
The tests are being triggered in jenkins.

FNALbuild commented:
This pull request is fully signed and it will be merged to develop and built in the next LArSoft release after it passes the integration tests. This pull request will now be reviewed by the release team before it's merged.)

FNALbuild commented:
+tests 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/8456&builds= 
 +build

FNALbuild commented:
+tests 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/8456&builds= 
 +build

ghost commented:
I will trigger another CI build since the DUNE build failed for unrelated reasons.

ghost commented:
trigger build

FNALbuild commented:
The tests are being triggered in jenkins.

FNALbuild commented:
+tests 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/8467&builds= 
 +build

FNALbuild commented:
+tests 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/8467&builds= 
 +build

ghost commented:
DUNE CI test looks good.

lgarren commented:
trigger build

FNALbuild commented:
The tests are being triggered in jenkins.

FNALbuild commented:
+tests 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/8518&builds= 
 +build

FNALbuild commented:
+tests 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/8518&builds= 
 +build

lgarren commented:
@ikatza   There is a c7 build problem.  Please fix.   CI tests are fine, but they do not build c7.

[ 90%] Building CXX object dunetpc/dune/OpticalDetector/CMakeFiles/ProtoDUNE_opdet_eventdisplay_module.dir/ProtoDUNE_opdet_eventdisplay_module.cc.o
/home/garren/scratch/larsoft/v08_52_00/srcs/larsim/larsim/LegacyLArG4/OpFastScintillation.cxx:1792:43: error: constexpr function never produces a constant expression [-Winvalid-constexpr]
  constexpr G4double OpFastScintillation::single_exp(const G4double t, const G4double tau2) const
                                          ^
/home/garren/scratch/larsoft/v08_52_00/srcs/larsim/larsim/LegacyLArG4/OpFastScintillation.cxx:1794:12: note: non-constexpr function 'exp' cannot be used in a constant expression
    return std::exp(-1.0 * t / tau2) / tau2;
           ^
/usr/include/bits/mathcalls.h:100:13: note: declared here
__MATHCALL (exp,, (_Mdouble_ __x));
            ^

ikatza commented:
Sure @lgarren. I guess clang doesn't allow constexpr for std::exp(). I used constexpr on a couple of functions with std::exp(), reckon it also failed there. Have removed it in both places.

Where can I see the full list of build errors for c7?

ikatza committed:
Remove constexpr qualifier from three functions to hanlde errors on clang builds

FNALbuild commented:
Pull request #14 was updated. @LArSoft/level-1-managers, @LArSoft/level-2-managers can you please check and sign again.

FNALbuild commented:
The code-checks are being triggered in jenkins.

FNALbuild commented:
+code-checks  
 for details see 
 https://buildmaster.fnal.gov/buildmaster/job/GithubIntegration/job/run-pr-code-checks//541/console

ikatza commented:
I removed the constexpr that used std::exp(). I did try to build on my Mac to investigate if there was more errors but couldn't, got the same error as is reported here: 
https://cdcvs.fnal.gov/redmine/issues/24337#note-4

lgarren commented:
@ikatza  There are still problems.  You can build with c7 on SLF.   For instance, on a gpvm server.

ikatza committed:
Remove some more constexpr to comply with clang

FNALbuild commented:
Pull request #14 was updated. @LArSoft/level-1-managers, @LArSoft/level-2-managers can you please check and sign again.

FNALbuild commented:
The code-checks are being triggered in jenkins.

FNALbuild commented:
+code-checks  
 for details see 
 https://buildmaster.fnal.gov/buildmaster/job/GithubIntegration/job/run-pr-code-checks//542/console

ikatza committed:
Revert "Use std::acos() in places where it's not called often". Turns out it's a hindrance and less consistent.

This reverts commit bf2530fb3af91f25440ca72f5aa09cb921a9ed58.

FNALbuild commented:
Pull request #14 was updated. @LArSoft/level-1-managers, @LArSoft/level-2-managers can you please check and sign again.

FNALbuild commented:
The code-checks are being triggered in jenkins.

FNALbuild commented:
+code-checks  
 for details see 
 https://buildmaster.fnal.gov/buildmaster/job/GithubIntegration/job/run-pr-code-checks//543/console

aihimmel commented:
Code changes look fine to me. On my phone so I can’t see the Jenkins result. If it builds it’s ready to go.


On May 13, 2020, at 5:21 AM, Fermilab build bot account <notifications@github.com> wrote:

﻿

+code-checks
for details see
https://buildmaster.fnal.gov/buildmaster/job/GithubIntegration/job/run-pr-code-checks//543/console

—
You are receiving this because you are on a team that was mentioned.
Reply to this email directly, view it on GitHub<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_LArSoft_larsim_pull_14-23issuecomment-2D627892202&d=DwMCaQ&c=gRgGjJ3BkIsb5y6s49QqsA&r=ILbuN39K86TRVV4M0fXV7A&m=0YXfKjh75d9rqCqXJ3U1riyAmNOG9Sw55BCj2lP9hXQ&s=78yLi7Aa7C5lqycM4FQIdswRea9aqVj3iqwcOe3AoWo&e=>, or unsubscribe<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_notifications_unsubscribe-2Dauth_AB345Q4EEU7JE5RZ5DBKUHLRRJYBPANCNFSM4MJWU6KQ&d=DwMCaQ&c=gRgGjJ3BkIsb5y6s49QqsA&r=ILbuN39K86TRVV4M0fXV7A&m=0YXfKjh75d9rqCqXJ3U1riyAmNOG9Sw55BCj2lP9hXQ&s=J1OtWyW4VEG-TSvPPLzcnh2TmF7kZExVw5E5W90Rxx4&e=>.

absolution1 commented:
trigger build

FNALbuild commented:
The tests are being triggered in jenkins.

FNALbuild commented:
+tests 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/8531&builds= 
 +build

FNALbuild commented:
+tests 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/8531&builds= 
 +build

absolution1 commented:
There's some kind of issue with accessing stash cache files in the SBND CI so the tests are delayed (ticket is getting submitted now).  Would everyone mind waiting until the problem is resolved so we can see the results of the SBND CI test?  I don't know if this issue is affecting other experiments as well.

ikatza commented:
Sure let's wait. My changes affect SBND, so it makes sense to make sure it works for it.

lgarren commented:
@absolution1   It looks like the SBND CI test finished successfully??  Are we good to go?  If not, this PR will have to wait until next week.

absolution1 commented:
All good, let's do this.

absolution1 commented:
approve

FNALbuild commented:
This pull request is fully signed and it will be merged to develop and built in the next LArSoft release after it passes the integration tests. This pull request will now be reviewed by the release team before it's merged.)

lgarren commented:
approve
